### DISCUSSION TITLE
Permission denied in new app

### DISCUSSION THREAD

**Discussion by mdh23rct** [idaholab/moose#25660] [Q&A Modules: Phase field]
Dear MOOSE experts,
I have developed a new app, compiled and built it successfully. The executable is generated after the "make -j 6" command. However, when I run the command "./run_tests -j 6", it shows "-bash: ./run_tests: Permission denied".

Would you be kind to let me know what is the problem? 
I am not getting it.

Thanks for your support.

**Comment by GiudGiud** [comment]
Hello

Something went wrong when you downloaded your code maybe. Did you `sudo git clone` ? that would cause that

the solution now is to change the permissions using the `chmod` command. Do at least that one, but maybe you ll need more

```
chmod +x run_tests
```

Guillaume

**Reply by mdh23rct** [reply]
Yes, first I cloned from my colleague's repository using Git desktop. then modified and built.
But, if I download the files as a Zip folder and work on it, this problem will be solved?

**Reply by GiudGiud** [reply]
git clone is the natural way to do this. Should work better than using zip files

**Reply by mdh23rct** [reply]
Dear Dr. Guillaume,
Having another problem:
/usr/bin/env: ‘python\r’: No such file or directory

while giving the command "./run_tests -j 6"

Is it indicating something? any solution for this problem?

Thanks.

**Reply by milljm** [reply]
I thought we adjusted all our shebangs with python3 instead of simply `python`. What file is this? Can you post the location of where it resides in the moose repo?

```bash
pwd
```

**Reply by mdh23rct** [reply]
Dear Dr. Jason Miller,
The problem is related to building an app from a repository using Git desktop.
I have successfull built the app after required modification. However when ./run_tests -j 6 command given, the error 
```
"/usr/bin/env: ‘python\r’: No such file or directory" 
```
pops up. Please be aware that the app is more than three years old. The only python file I am seeing in my application folder (/home/user/projects/test3/melt/doc):
```
#* Licensed under LGPL 2.1, please see LICENSE for details
#* https://www.gnu.org/licenses/lgpl-2.1.html

import sys
import os

# Locate MOOSE directory
MOOSE_DIR = os.getenv('MOOSE_DIR', os.path.abspath(os.path.join(os.path.dirname(__name__), '..', 'moose')))
if not os.path.exists(MOOSE_DIR):
    MOOSE_DIR = os.path.abspath(os.path.join(os.path.dirname(__name__), '..', '..', 'moose'))
if not os.path.exists(MOOSE_DIR):
    raise Exception('Failed to locate MOOSE, specify the MOOSE_DIR environment variable.')
os.environ['MOOSE_DIR'] = MOOSE_DIR

# Append MOOSE python directory
MOOSE_PYTHON_DIR = os.path.join(MOOSE_DIR, 'python')
if MOOSE_PYTHON_DIR not in sys.path:
    sys.path.append(MOOSE_PYTHON_DIR)

from MooseDocs import main
if __name__ == '__main__':
    sys.exit(main.run()) """
```
So what modification I should do and in which file to solve this python problem, would you be kind to let me know!
I would appreciate your advice.

Thanks and best regards.

**Reply by GiudGiud** [reply]
add this at the top of any python script you use
```
#!/usr/bin/env python3
```

it s because your app is old and not updated that you're not having this I think

**Reply by milljm** [reply]
I am not sure that will work. But worth a shot. I see no SHEBANG set at all... so I am not even sure where it though to try and run `/usr/bin/env python`.

**Reply by GiudGiud** [reply]
we have `#!/usr/bin/env python3` at the top of run_tests too

**Reply by milljm** [reply]
I am talking about what was posted as @mdh23rct's version of `run_tests`. I just don't see any SHEBANG there. So I am unsure what file we need to alter. Unless it was just a bad copy/paste?

**Reply by mdh23rct** [reply]
Dear Dr. Guillaume and Dr. Miller,
I have added the SHEBANG in moosedocs.py and run_tests. However, no luck. 
Then, I checked the data format of the files, since I am using wsl in windows. 
I have converted the run_tests from dos2unix format. This time ./run_tests -j 6 worked, but test of all objects failed.
Maybe I have some data compatibility issues! You have any comment on that!

Best regards.

**Reply by GiudGiud** [reply]
> , but test of all objects failed.

what was the full message for one of them?

**Reply by mdh23rct** [reply]
Here, I am attaching the error message. Pretty long:
```
"""test:kernels/surface_reaction.1_reaction_kernel .............................................. FAILED (ERRMSG)
test:kernels/simple_diffusion.test: Working Directory: /home/user/projects/plasma/test/tests/kernels/simple_diffusion
test:kernels/simple_diffusion.test: Running command: /home/user/projects/plasma/plasma-opt -i simple_diffusion.i --no-gdb-backtrace
test:kernels/simple_diffusion.test:
test:kernels/simple_diffusion.test:
test:kernels/simple_diffusion.test: *** Warning, This code is deprecated and will be removed in future versions:
test:kernels/simple_diffusion.test: "registerAll" in Modules is deprecated. Please update your *App.C file(s) to call the new templated "registerAllObjects" method (e.g. ModulesApp::registerAllObjects<MyApp>(...))
test:kernels/simple_diffusion.test: Stack frames: 14
test:kernels/simple_diffusion.test: 0: libMesh::print_trace(std::ostream&)
test:kernels/simple_diffusion.test: 1: void moose::internal::mooseDeprecatedStream<libMesh::BasicOStreamProxy<char, std::char_traits<char> >, char const (&) [178]>(libMesh::BasicOStreamProxy<char, std::char_traits<char> >&, bool, bool, char const (&) [178])
test:kernels/simple_diffusion.test: 2: ModulesApp::registerAll(Factory&, ActionFactory&, Syntax&)
test:kernels/simple_diffusion.test: 3: plasmaApp::registerAll(Factory&, ActionFactory&, Syntax&)
test:kernels/simple_diffusion.test: 4: plasmaTestApp::registerAll(Factory&, ActionFactory&, Syntax&, bool)
test:kernels/simple_diffusion.test: 5: plasmaTestApp::plasmaTestApp(InputParameters)
test:kernels/simple_diffusion.test: 6: /home/user/projects/plasma/test/lib/libplasma_test-opt.so.0(+0xb1e8) [0x7f0ae2cbc1e8]
test:kernels/simple_diffusion.test: 7: /home/user/projects/plasma/test/lib/libplasma_test-opt.so.0(+0xb277) [0x7f0ae2cbc277]
test:kernels/simple_diffusion.test: 8: AppFactory::createShared(std::__cxx11::basic_string<char, std::char_traits<char>, std::allocator<char> > const&, std::__cxx11::basic_string<char, std::char_traits<char>, std::allocator<char> > const&, InputParameters, int)
test:kernels/simple_diffusion.test: 9: AppFactory::createAppShared(std::__cxx11::basic_string<char, std::char_traits<char>, std::allocator<char> > const&, int, char**, int)
test:kernels/simple_diffusion.test: 10: main
test:kernels/simple_diffusion.test: 11: /lib/x86_64-linux-gnu/libc.so.6(+0x29d90) [0x7f0ad4728d90]
test:kernels/simple_diffusion.test: 12: __libc_start_main
test:kernels/simple_diffusion.test: 13: /home/user/projects/plasma/plasma-opt(+0x33ef) [0x55c556a6a3ef]
test:kernels/simple_diffusion.test:
test:kernels/simple_diffusion.test:
test:kernels/simple_diffusion.test:
test:kernels/simple_diffusion.test: *** ERROR ***
test:kernels/simple_diffusion.test: /home/user/projects/plasma/test/tests/kernels/simple_diffusion/simple_diffusion.i contains Windows(DOS) line endings which are not supported.
test:kernels/simple_diffusion.test:
test:kernels/simple_diffusion.test: Stack frames: 11
test:kernels/simple_diffusion.test: 0: libMesh::print_trace(std::ostream&)
test:kernels/simple_diffusion.test: 1: moose::internal::mooseErrorRaw(std::__cxx11::basic_string<char, std::char_traits<char>, std::allocator<char> >, std::__cxx11::basic_string<char, std::char_traits<char>, std::allocator<char> >)
test:kernels/simple_diffusion.test: 2: void mooseError<std::__cxx11::basic_string<char, std::char_traits<char>, std::allocator<char> > >(std::__cxx11::basic_string<char, std::char_traits<char>, std::allocator<char> >&&)
test:kernels/simple_diffusion.test: 3: MooseUtils::checkFileReadable(std::__cxx11::basic_string<char, std::char_traits<char>, std::allocator<char> > const&, bool, bool, bool)
test:kernels/simple_diffusion.test: 4: Parser::parse(std::vector<std::__cxx11::basic_string<char, std::char_traits<char>, std::allocator<char> >, std::allocator<std::__cxx11::basic_string<char, std::char_traits<char>, std::allocator<char> > > > const&, std::__cxx11::basic_string<char, std::char_traits<char>, std::allocator<char> > const&)
test:kernels/simple_diffusion.test: 5: MooseApp::setupOptions()
test:kernels/simple_diffusion.test: 6: MooseApp::run()
test:kernels/simple_diffusion.test: 7: main
test:kernels/simple_diffusion.test: 8: /lib/x86_64-linux-gnu/libc.so.6(+0x29d90) [0x7f0ad4728d90]
test:kernels/simple_diffusion.test: 9: __libc_start_main
test:kernels/simple_diffusion.test: 10: /home/user/projects/plasma/plasma-opt(+0x33ef) [0x55c556a6a3ef]
test:kernels/simple_diffusion.test: application called MPI_Abort(MPI_COMM_WORLD, 1) - process 0
test:kernels/simple_diffusion.test: [unset]: write_line error; fd=-1 buf=:cmd=abort exitcode=1
test:kernels/simple_diffusion.test: :
test:kernels/simple_diffusion.test: system msg for write_line failure : Bad file descriptor
test:kernels/simple_diffusion.test:
test:kernels/simple_diffusion.test:
test:kernels/simple_diffusion.test: *** Warning, This code is deprecated and will be removed in future versions:
test:kernels/simple_diffusion.test: "registerAll" in Modules is deprecated. Please update your *App.C file(s) to call the new templated "registerAllObjects" method (e.g. ModulesApp::registerAllObjects<MyApp>(...))
test:kernels/simple_diffusion.test: Stack frames: 14
test:kernels/simple_diffusion.test: 0: libMesh::print_trace(std::ostream&)
test:kernels/simple_diffusion.test: 1: void moose::internal::mooseDeprecatedStream<libMesh::BasicOStreamProxy<char, std::char_traits<char> >, char const (&) [178]>(libMesh::BasicOStreamProxy<char, std::char_traits<char> >&, bool, bool, char const (&) [178])
test:kernels/simple_diffusion.test: 2: ModulesApp::registerAll(Factory&, ActionFactory&, Syntax&)
test:kernels/simple_diffusion.test: 3: plasmaApp::registerAll(Factory&, ActionFactory&, Syntax&)
test:kernels/simple_diffusion.test: 4: plasmaTestApp::registerAll(Factory&, ActionFactory&, Syntax&, bool)
test:kernels/simple_diffusion.test: 5: plasmaTestApp::plasmaTestApp(InputParameters)
test:kernels/simple_diffusion.test: 6: /home/user/projects/plasma/test/lib/libplasma_test-opt.so.0(+0xb1e8) [0x7f0ae2cbc1e8]
test:kernels/simple_diffusion.test: 7: /home/user/projects/plasma/test/lib/libplasma_test-opt.so.0(+0xb277) [0x7f0ae2cbc277]
test:kernels/simple_diffusion.test: 8: AppFactory::createShared(std::__cxx11::basic_string<char, std::char_traits<char>, std::allocator<char> > const&, std::__cxx11::basic_string<char, std::char_traits<char>, std::allocator<char> > const&, InputParameters, int)
test:kernels/simple_diffusion.test: 9: AppFactory::createAppShared(std::__cxx11::basic_string<char, std::char_traits<char>, std::allocator<char> > const&, int, char**, int)
test:kernels/simple_diffusion.test: 10: main
test:kernels/simple_diffusion.test: 11: /lib/x86_64-linux-gnu/libc.so.6(+0x29d90) [0x7f0ad4728d90]
test:kernels/simple_diffusion.test: 12: __libc_start_main
test:kernels/simple_diffusion.test: 13: /home/user/projects/plasma/plasma-opt(+0x33ef) [0x55c556a6a3ef]
test:kernels/simple_diffusion.test:
test:kernels/simple_diffusion.test:
test:kernels/simple_diffusion.test:
test:kernels/simple_diffusion.test: *** ERROR ***
test:kernels/simple_diffusion.test: /home/user/projects/plasma/test/tests/kernels/simple_diffusion/simple_diffusion.i contains Windows(DOS) line endings which are not supported.
test:kernels/simple_diffusion.test:
test:kernels/simple_diffusion.test: Stack frames: 11
test:kernels/simple_diffusion.test: 0: libMesh::print_trace(std::ostream&)
test:kernels/simple_diffusion.test: 1: moose::internal::mooseErrorRaw(std::__cxx11::basic_string<char, std::char_traits<char>, std::allocator<char> >, std::__cxx11::basic_string<char, std::char_traits<char>, std::allocator<char> >)
test:kernels/simple_diffusion.test: 2: void mooseError<std::__cxx11::basic_string<char, std::char_traits<char>, std::allocator<char> > >(std::__cxx11::basic_string<char, std::char_traits<char>, std::allocator<char> >&&)
test:kernels/simple_diffusion.test: 3: MooseUtils::checkFileReadable(std::__cxx11::basic_string<char, std::char_traits<char>, std::allocator<char> > const&, bool, bool, bool)
test:kernels/simple_diffusion.test: 4: Parser::parse(std::vector<std::__cxx11::basic_string<char, std::char_traits<char>, std::allocator<char> >, std::allocator<std::__cxx11::basic_string<char, std::char_traits<char>, std::allocator<char> > > > const&, std::__cxx11::basic_string<char, std::char_traits<char>, std::allocator<char> > const&)
test:kernels/simple_diffusion.test: 5: MooseApp::setupOptions()
test:kernels/simple_diffusion.test: 6: MooseApp::run()
test:kernels/simple_diffusion.test: 7: main
test:kernels/simple_diffusion.test: 8: /lib/x86_64-linux-gnu/libc.so.6(+0x29d90) [0x7f0ad4728d90]
test:kernels/simple_diffusion.test: 9: __libc_start_main
test:kernels/simple_diffusion.test: 10: /home/user/projects/plasma/plasma-opt(+0x33ef) [0x55c556a6a3ef]
test:kernels/simple_diffusion.test: application called MPI_Abort(MPI_COMM_WORLD, 1) - process 0
test:kernels/simple_diffusion.test: [unset]: write_line error; fd=-1 buf=:cmd=abort exitcode=1
test:kernels/simple_diffusion.test: :
test:kernels/simple_diffusion.test: system msg for write_line failure : Bad file descriptor
test:kernels/simple_diffusion.test:
test:kernels/simple_diffusion.test:
test:kernels/simple_diffusion.test: Exit Code: 1
test:kernels/simple_diffusion.test: ################################################################################
test:kernels/simple_diffusion.test: Tester failed, reason: ERRMSG
test:kernels/simple_diffusion.test:
test:kernels/simple_diffusion.test ........................................................... FAILED (ERRMSG)"""
```

I have filtered out some serious issues with the code:
1.
test:kernels/simple_diffusion.test: [unset]: write_line error; fd=-1 buf=:cmd=abort exitcode=1
test:kernels/simple_diffusion.test: :
test:kernels/simple_diffusion.test: system msg for write_line failure : Bad file descriptor

2.
test:kernels/simple_diffusion.test: application called MPI_Abort(MPI_COMM_WORLD, 1) - process 0
test:kernels/simple_diffusion.test: [unset]: write_line error; fd=-1 buf=:cmd=abort exitcode=1

test:kernels/simple_diffusion.test: :
test:kernels/simple_diffusion.test: system msg for write_line failure : Bad file descriptor

##What this Bad file descriptor indicates?? Any specific steps to take to solve it??

3.
test:kernels/simple_diffusion.test: /home/user/projects/plasma/test/tests/kernels/simple_diffusion/simple_diffusion.i contains Windows(DOS) line endings which are not supported.

##I have to save the file in the wsl format. I believe, this particular error will be solved then.

You have commens on the other errors existing in this file??

Thanks for your support.

**Reply by GiudGiud** [reply]
3 is the thing you need to address.

The other things are just consequences of us throwing an error because of the DOS line endings.
You need to avoid opening these files in windows and saving them there

Now to convert a file back, you can use the `dos2unix` command

**Reply by mdh23rct** [reply]
Dear Dr. Guillaume, 
Luckily now the errors came down from five to two. The remaining two errors are:

1. test:materials/fiber_direction.1_direction_calculation: Running exodiff: /home/user/projects/moose/framework/contrib/exodiff/exodiff -m  -F 1e-10 -t 5.5e-06  /home/user/projects/plasma/test/tests/materials/fiber_direction/gold/fiber_direction_exodus.e /home/user/projects/plasma/test/tests/materials/fiber_direction/fiber_direction_exodus.e
test:materials/fiber_direction.1_direction_calculation: ERROR:
test:materials/fiber_direction.1_direction_calculation:    *****************************************************************
test:materials/fiber_direction.1_direction_calculation:               EXODIFF   (Version: 2.90) Modified: 2018-02-15
test:materials/fiber_direction.1_direction_calculation:               Authors:  Richard Drake, rrdrake@sandia.gov
test:materials/fiber_direction.1_direction_calculation:                         Greg Sjaardema, gdsjaar@sandia.gov
test:materials/fiber_direction.1_direction_calculation:               Run on    2023/10/04   19:33:55 BST
test:materials/fiber_direction.1_direction_calculation:    *****************************************************************

2. test:kernels/plasma_reaction.1_reaction_kernel: The following error occurred in the object "MOOSE Problem", of type "FEProblem".
test:kernels/plasma_reaction.1_reaction_kernel:
test:kernels/plasma.1_reaction_kernel: Material property 'Ref', requested by 'omega_b_face' is not defined on block 0.

Would you be kind to advice me on how to remove these errors? Specially, I am not understanding what is happening with the Exodiff test!

Thanks.

**Reply by GiudGiud** [reply]
Not sure.

There isn't enough information either to debug them. 

It looks the input file isn't working because some material property is not defined

And these tests look like they are part of your application not MOOSE? So it s not a given that they were working 3 years ago. 

I would move on. Please make sure to push your fixes online to make sure the next person using this application doesn't have to do the same steps

**Reply by mdh23rct** [reply]
Dear Dr. Guillaume,
Thanks for your time and support. Just one last question. I looked at the MOOSE documentation regarding Exodiff. It compares Exodus output(s) for the test to a reference in the specified gold_dir folder. That means I can turn off the Exodus test in the application if I like. Is it true?

And off course as you have asked, I am determined to update the existing repository as soon as I have the error-free files.

Again I appreciate your support and time.

**Reply by milljm** [reply]
There exist the ability to "execute only" and not perform the post process portion of the test. Which is what you are asking. But if you do that, you might as well just "move on" if you are not interested in test results.

Edit: Actually I am not seeing that option. I thought we had one. The option I was thinking of runs the post process step, and not the application (should_execute = false).

Instead, if you wanted to see your failing test pass, you would have to change that test type to 'RunApp' instead of 'Exodiff'.

**Reply by mdh23rct** [reply]
Dear Dr. Guillaume and Dr. Miller,
Happy that you helped me out. Finally, I have been able to build the repository without any error.
Thanks for your time and support.
Lets see next how the file execution goes on.

Best regards.
